subsys/bluetooth/host/audio/iso.c

```diff
@@ -0,0 +1,1075 @@
+		if (chan->ops->disconnected) {
+			chan->ops->disconnected(chan);
+		}
+
+		chan->conn = NULL;
+		bt_iso_chan_set_state(chan, BT_ISO_DISCONNECTED);
+	}
+}
+
+int bt_iso_server_register(struct bt_iso_server *server)
+{
+	__ASSERT_NO_MSG(server);
+
+	if (iso_server) {
+		return -EADDRINUSE;
+	}
+
+	if (!server->accept) {
+		return -EINVAL;
+	}
+
+	if (server->sec_level > BT_SECURITY_L3) {
+		return -EINVAL;
+	} else if (server->sec_level < BT_SECURITY_L1) {
+		/* Level 0 is only applicable for BR/EDR */
+		server->sec_level = BT_SECURITY_L1;
+	}
+
+	BT_DBG("%p", server);
+
+	iso_server = server;
+
+	return 0;
+}
+
+#if defined(CONFIG_BT_AUDIO_DEBUG_ISO)
+const char *bt_iso_chan_state_str(bt_iso_chan_state_t state)
+{
+	switch (state) {
+	case BT_ISO_DISCONNECTED:
+		return "disconnected";
+	case BT_ISO_BOUND:
+		return "bound";
+	case BT_ISO_CONNECT:
+		return "connect";
+	case BT_ISO_CONNECTED:
+		return "connected";
+	case BT_ISO_DISCONNECT:
+		return "disconnect";
+	default:
+		return "unknown";
+	}
+}
+
+void bt_iso_chan_set_state_debug(struct bt_iso_chan *chan,
+				 bt_iso_chan_state_t state,
+				 const char *func, int line)
+{
+	BT_DBG("chan %p conn %p %s -> %s", chan, chan->conn,
+	       bt_iso_chan_state_str(chan->state),
+	       bt_iso_chan_state_str(state));
+
+	/* check transitions validness */
+	switch (state) {
+	case BT_ISO_DISCONNECTED:
+		/* regardless of old state always allows this state */
+		break;
+	case BT_ISO_BOUND:
+		if (chan->state != BT_ISO_DISCONNECTED) {
+			BT_WARN("%s()%d: invalid transition", func, line);
+		}
+		break;
+	case BT_ISO_CONNECT:
+		if (chan->state != BT_ISO_BOUND) {
+			BT_WARN("%s()%d: invalid transition", func, line);
+		}
+		break;
+	case BT_ISO_CONNECTED:
+		if (chan->state != BT_ISO_BOUND &&
+		    chan->state != BT_ISO_CONNECT) {
+			BT_WARN("%s()%d: invalid transition", func, line);
+		}
+		break;
+	case BT_ISO_DISCONNECT:
+		if (chan->state != BT_ISO_CONNECTED) {
+			BT_WARN("%s()%d: invalid transition", func, line);
+		}
+		break;
+	default:
+		BT_ERR("%s()%d: unknown (%u) state was set", func, line, state);
+		return;
+	}
+
+	chan->state = state;
+}
+#else
+void bt_iso_chan_set_state(struct bt_iso_chan *chan, uint8_t state)
+{
+	chan->state = state;
+}
+#endif /* CONFIG_BT_AUDIO_DEBUG_ISO */
+
+void bt_iso_chan_remove(struct bt_conn *conn, struct bt_iso_chan *chan)
+{
+	struct bt_iso_chan *c;
+	sys_snode_t *prev = NULL;
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&conn->channels, c, node) {
+		if (c == chan) {
+			sys_slist_remove(&conn->channels, prev, &chan->node);
+			return;
+		}
+
+		prev = &chan->node;
+	}
+}
+
+int bt_iso_chan_bind(struct bt_conn **conns, uint8_t num_conns,
+		     struct bt_iso_chan **chans)
+{
+	struct bt_iso_create_param param;
+	int i, err;
+
+	__ASSERT_NO_MSG(conns);
+	__ASSERT_NO_MSG(num_conns);
+	__ASSERT_NO_MSG(chans);
+
+	memset(&param, 0, sizeof(param));
+
+	/* TODO: Allow setting ID? */
+	param.id = 0x01;
```

Comment:
Not sure if the API should allow setting the ID, but shouldn't the ID at least be incremented for each call, or something?

Reply:
It should, or actually have it allocated with connection index or something like it, I will fix it.